src/main/java/com/google/devtools/build/lib/rules/cpp/Cpp20ModuleDepMapAction.java

```diff
@@ -0,0 +1,222 @@
+ * 1. the modules needed for compiling C++ Modules or ordinary sources
+ * are provided by the ddi file
+ * 2. all modules information provided by the modules info file
+ * 3. the output file .modmap collect all modules, including indirect dependencies
+ * e.g. a -> b -> c -> d
+ * the ddi file only record module a require module b
+ * However, in .modmap file, we need to put all modules (b, c, d) in it
+ * the format of .modmap file is
+ * -fmodule-file=<module-name1>=<module-path1>
+ * -fmodule-file=<module-name2>=<module-path2>
+ * ...
+ * 4. the output file .modmap.input collects all module files' paths
+ * it is convenient for CppCompileAction to handle dynamic modules input with the .modmap.input file
+ * otherwise, we need to parse the .modmap file and get all modules' path in CppCompileAction
+ * the format of .modmap file is
+ * <module-path1>
+ * <module-path2>
+ * ...
+ */
+@Immutable
+public final class Cpp20ModuleDepMapAction extends AbstractAction {
+  private static final String GUID = "b02b043a-6c4e-4eeb-992a-09769b366fdd";
+  Artifact ddiFile;
+  Artifact modulesInfoFile;
+  Artifact modmapFile;
+  Artifact modmapInputFile;
+
+  public Cpp20ModuleDepMapAction(
+      ActionOwner owner,
+      Artifact ddiFile,
+      Artifact modulesInfoFile,
+      Artifact modmapFile,
+      Artifact modmapInputFile) {
+    super(
+        owner,
+        NestedSetBuilder.<Artifact>stableOrder().add(ddiFile).add(modulesInfoFile).build(),
+        ImmutableSet.of(modmapFile, modmapInputFile));
+    this.ddiFile = ddiFile;
+    this.modulesInfoFile = modulesInfoFile;
+    this.modmapFile = modmapFile;
+    this.modmapInputFile = modmapInputFile;
+  }
+
+  @Override
+  public ActionResult execute(ActionExecutionContext actionExecutionContext)
+      throws ActionExecutionException, InterruptedException {
+    try {
+      Map<String, String> moduleMap = computeModuleMap(actionExecutionContext);
+
+      var result1 =
+          writeOutputToFile(
+              actionExecutionContext,
+              out -> {
+                OutputStreamWriter content =
+                    new OutputStreamWriter(out, StandardCharsets.ISO_8859_1);
+                for (Map.Entry<String, String> entry : moduleMap.entrySet()) {
+                  String moduleName = entry.getKey();
+                  String modulePath = entry.getValue();
+                  content.append("-f");
+                  content.append("module-file=");
+                  content.append(moduleName);
+                  content.append("=");
+                  content.append(modulePath);
+                  content.append('\n');
+                }
+                content.flush();
+              },
+              modmapFile);
+      var result2 =
+          writeOutputToFile(
+              actionExecutionContext,
+              out -> {
+                OutputStreamWriter content =
+                    new OutputStreamWriter(out, StandardCharsets.ISO_8859_1);
```

Comment:
Woah; not UTF-8? Or are paths in Bazel essentially limited to ASCII? `Charset.defaultCharset()` seems better?